core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,214 @@
+  def nodeBlacklist(): Set[String] = {
+    _nodeBlacklist.get()
+  }
+
+  def isNodeBlacklisted(node: String): Boolean = {
+    nodeIdToBlacklistExpiryTime.contains(node)
+  }
+
+  def removeExecutor(executorId: String): Unit = {
+    // we intentionally do not clean up executors that are already blacklisted, so that if another
+    // executor on the same node gets blacklisted, we can blacklist the entire node.
+    executorIdToFailureCount -= executorId
+  }
+}
+
+
+private[scheduler] object BlacklistTracker extends Logging {
+
+  /**
+   * Return true if the blacklist is enabled, based on the following order of preferences:
+   * 1. Is it specifically enabled or disabled?
+   * 2. Is it enabled via the legacy timeout conf?
+   * 3. Use the default for the spark-master:
+   *   - off for local mode
+   *   - on for distributed modes (including local-cluster)
+   */
+  def isBlacklistEnabled(conf: SparkConf): Boolean = {
+    conf.get(config.BLACKLIST_ENABLED) match {
+      case Some(isEnabled) =>
+        isEnabled
+      case None =>
+        // if they've got a non-zero setting for the legacy conf, always enable the blacklist,
+        // otherwise, use the default based on the cluster-mode (off for local-mode, on otherwise).
+        val legacyKey = config.BLACKLIST_LEGACY_TIMEOUT_CONF.key
+        conf.get(config.BLACKLIST_LEGACY_TIMEOUT_CONF) match {
+          case Some(legacyTimeout) if legacyTimeout == 0 =>
+            logWarning(s"Turning off blacklisting due to legacy configuaration:" +
+              s" $legacyKey == 0")
+            false
+          case Some(legacyTimeout) =>
+              // mostly this is necessary just for tests, since real users that want the blacklist
+              // will get it anyway by default
+              logWarning(s"Turning on blacklisting due to legacy configuration:" +
+                s" $legacyKey > 0")
+              true
+          case None =>
+            // local-cluster is *not* considered local for these purposes, we still want the
+            // blacklist enabled by default
+            !Utils.isLocalMaster(conf)
+        }
+    }
+  }
+
+  def getBlacklistExpiryTime(conf: SparkConf): Long = {
+    val timeoutConf = conf.get(config.BLACKLIST_EXPIRY_TIMEOUT_CONF)
+    val legacyTimeoutConf = conf.get(config.BLACKLIST_LEGACY_TIMEOUT_CONF)
+    (timeoutConf, legacyTimeoutConf) match {
+      case (Some(x), _) => x
+      case (None, Some(y)) => y
+      case (None, None) =>
+        Utils.timeStringAsMs("1h")
+    }
+  }
+}
+
+/** Failures for one executor, within one taskset */
+private[scheduler] final class FailureStatus(val node: String) {
+  /** index of the tasks in the taskset that have failed on this executor. */
+  val tasksWithFailures = HashSet[Int]()
+  var totalFailures = 0
+
+  override def toString(): String = {
+    s"totalFailures = $totalFailures; tasksFailed = $tasksWithFailures"
+  }
+}
+
+private final case class BlacklistedExecutor(node: String, expiryTime: Long)
```

Comment:
how about BlacklistStatus? "Executor" is confusing, since the class doesn't contain anything about an executor